keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,213 @@
+- https://github.com/kubernetes/community/pull/278
+- https://github.com/kubernetes/community/pull/620
+- https://github.com/kubernetes/kubernetes/pull/44597
+- https://github.com/kubernetes/kubernetes/pull/50296
+- https://github.com/kubernetes/kubernetes/pull/70436
+
+Server-side [apply](http://features.k8s.io/555) is what enables this proposal to
+become possible.
+
+### Goals
+
+The goal is to enable a union or "oneof" semantic in Kubernetes types, both for
+in-tree types and for CRDs.
+
+### Non-Goals
+
+We're not planning to use this KEP to release the feature, but mostly as a way
+to document what we're doing.
+
+## Proposal
+
+In order to support unions in a backward compatible way in kubernetes, we're
+proposing the following changes.
+
+Note that this proposes unions to be "at most one of". Whether exactly one is
+supported or not should be implemented by the validation logic.
+
+### Go tags
+
+We're proposing a new type of tags for go types (in-tree types, and also
+kubebuilder types):
+
+- `// +union` before a field means that it is part of the union. Multiple fields
+  can have this prefix. These fields MUST BE optional, omitempty and pointers.
+- `// +discriminator` before a field means that this field is the discriminator
+  for the union. Only one field per structure can have this prefix. This field HAS
+  TO be a string, and CAN be optional.
+
+Note that this means that only one "union" type can be defined per structure,
+which is a deliberate decision. Go embedding which, once flattened, would define
+multiple unions in the same structure are also forbidden. Also note that we have
+no occurrences of these existing in core-types (AFAIK).
+
+### OpenAPI
+
+OpenAPI v3 already allows a "oneOf" form, which is accepted by CRD validation
+(and will continue to be accepted in the future). That oneOf form will be used
+for validation, but is "on-top" of this proposal.
+
+New extensions are created in the openapi to describe the behavior:
+- `x-kubernetes-union-discriminator: <discriminator>` is set to the name of the
+  discriminator field, if present,
+- `x-kubernetes-union-fields-discriminated: {"<fieldName>":
+  "<discriminatedName>"}` is a map of fields that belong to the union to their
+  discriminated names. The discriminatedValue will typically be set to the name
+  of the Go variable.
+
+These fields belong directly in the object that holds the union.
+
+Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.
+
+As for Go types, we only accept one union per structure (including embeddeing),
+which means there can be only one discriminator, and all other `oneof` fields
+are assumed to belong to the same union.
+
+Only one union is allowed per openapi objects (even after flattening of allOfs).
+
+### Discriminator
+
+For backward compatibility reasons, discriminators can be added to existing
+union structures as an optional string. This has a nice property that it's going
```

Comment:
If this value is defaulted, it does not have to be optional anymore.

Reply:
which remind me that client side validation does not see defaults yet and hence falls over a non-optional discriminator. I guess that's what you have in mind.

Reply:
Right, if we make it required, then client-side validation will force people to set it, which we don't want.

Reply:
On one hand, it should be OK to have clients be aware of defaulting rules (they are part of the API), so clients could make a copy, apply defaults, and validate that.

On the other hand, that fails in case of admission controllers.  But if that is a requirement, then client-side validation can never work.

I know this is a little controversial, but I think it's OK to client-side default (in a temporary copy!!) to validate.

Reply:
we have dry-run, don't we?

Reply:
Sure

Reply:
We could also just stop validating required fields client-side.

Reply:
I don't see why this is important anyway, you can have a look at the behavior here (which I think is a big improvement): https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/union_test.go